crates/signer-gcp/src/signer.rs

```diff
@@ -0,0 +1,348 @@
+/// Note that this signer only supports asynchronous operations. Calling a non-asynchronous method
+/// will always return an error.
+///
+/// # Examples
+///
+/// ```no_run
+/// use alloy_signer::Signer;
+/// use alloy_signer_gcp::{GcpKeyRingRef, GcpSigner};
+/// use gcloud_sdk::{
+///     google::cloud::kms::v1::key_management_service_client::KeyManagementServiceClient,
+///     GoogleApi,
+/// };
+///
+/// # async fn test() {
+///
+/// let project_id = std::env::var("GOOGLE_PROJECT_ID").expect("GOOGLE_PROJECT_ID");
+/// let location = std::env::var("GOOGLE_LOCATION").expect("GOOGLE_LOCATION");
+/// let keyring_name = std::env::var("GOOGLE_KEYRING").expect("GOOGLE_KEYRING");
+///
+/// let keyring = GcpKeyRingRef::new(&project_id, &location, &keyring_name);
+/// let client = GoogleApi::from_function(
+///     KeyManagementServiceClient::new,
+///     "https://cloudkms.googleapis.com",
+///     None,
+/// )
+/// .await
+/// .expect("Failed to create GCP KMS Client");
+///
+/// let key_id = "..".to_string();
+/// let key_version = 1;
+/// let chain_id = 1;
+/// let signer = GcpSigner::new(client, keyring, key_id, key_version, chain_id).await.unwrap();
+///
+/// let message = vec![0, 1, 2, 3];
+///
+/// let sig = signer.sign_message(&message).await.unwrap();
+/// assert_eq!(sig.recover_address_from_msg(message).unwrap(), signer.address());
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct GcpSigner {
+    client: Client,
+    keyring_ref: GcpKeyRingRef,
+    chain_id: u64,
+    key_version: u64,
+    key_id: String,
+    pubkey: VerifyingKey,
+    address: Address,
+}
+
+impl fmt::Debug for GcpSigner {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("GcpSigner")
+            .field("key_id", &self.key_id)
+            .field("key_version", &self.key_version)
+            .field("chain_id", &self.chain_id)
+            .field("pubkey", &hex::encode(self.pubkey.to_sec1_bytes()))
+            .field("address", &self.address)
+            .finish()
+    }
+}
+
+/// Errors thrown by [`GcpSigner`].
+#[derive(Error, Debug)]
+pub enum GcpSignerError {
+    /// Thrown when the GCP KMS API returns a signing error.
+    #[error(transparent)]
+    GoogleKmsError(#[from] gcloud_sdk::error::Error),
+
+    /// Thrown on a request error.
+    #[error(transparent)]
+    RequestError(#[from] tonic::Status),
+
+    /// [`spki`] error.
+    #[error(transparent)]
+    Spki(#[from] spki::Error),
+
+    /// [`ecdsa`] error.
+    #[error(transparent)]
+    K256(#[from] ecdsa::Error),
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl Signer for GcpSigner {
+    #[instrument(err)]
+    #[allow(clippy::blocks_in_conditions)]
+    async fn sign_hash(&self, hash: &B256) -> Result<Signature> {
+        self.sign_digest_with_eip155(hash, self.chain_id).await.map_err(alloy_signer::Error::other)
+    }
+
+    #[cfg(TODO)] // TODO: TypedTransaction
+    #[instrument(err)]
```

Comment:
#83 has the latest work on this. We should merge this PR in advance of #83, and then update this code to match that code